tensorflow_quantum/core/ops/batch_util.py

```diff
@@ -451,45 +428,30 @@ def batch_calculate_expectation(circuits, param_resolvers, ops, simulator):
                 raise TypeError('ops must contain only cirq.PauliSum objects.'
                                 ' Given: {}'.format(type(x)))
 
-    return_mem_shape = (len(circuits), len(ops[0]))
-    if isinstance(simulator, cirq.DensityMatrixSimulator):
-        post_process = lambda op, state, order: sum(
-            x._expectation_from_density_matrix_no_validation(
-                state.final_density_matrix, order) for x in op).real
-    elif isinstance(simulator, cirq.Simulator):
-        post_process = \
-            lambda op, state, order: op.expectation_from_state_vector(
-                state.final_state_vector, order).real
-    else:
-        raise TypeError('Simulator {} is not supported by '
-                        'batch_calculate_expectation.'.format(type(simulator)))
-
-    shared_array = _make_simple_view(return_mem_shape, -2, np.float32, 'f')
-
-    # avoid mutating ops array
-    ops = np.copy(ops)
-    # TODO (mbbrough): make cirq PauliSUms pickable at some point ?
-    for i in range(len(ops)):
-        for j in range(len(ops[i])):
-            ops[i][j] = serializer.serialize_paulisum(ops[i][j])
-
-    input_args = list(
-        _prep_pool_input_args(list(
-            itertools.product(range(len(circuits)), range(len(ops[0])))),
-                              circuits,
-                              param_resolvers,
-                              ops,
-                              slice_args=False))
-
-    with ProcessPool(processes=None,
-                     initializer=_setup_dict,
-                     initargs=(shared_array, return_mem_shape, simulator,
-                               post_process)) as pool:
-
-        pool.starmap(_analytical_expectation_worker_func, input_args)
-
-    return _convert_simple_view_to_result(shared_array, np.float32,
-                                          return_mem_shape)
+    all_exp_vals = []
+    for c, p, o in zip(circuits, param_resolvers, ops):
+        # Convention in TFQ is to set expectations of empty circuits to -2.
+        if len(c) == 0:
+            all_exp_vals.append(np.full(len(o), -2.0, dtype=np.float32))
+        else:
+            # TODO(zaqqwerty): remove DM sim check once cirq #3964 is resolved.
+            if isinstance(simulator, cirq.DensityMatrixSimulator):
+                qubit_order = dict(
+                    zip(sorted(c.all_qubits()),
+                        list(range(len(c.all_qubits())))))
+                sim_result = simulator.simulate(c, p)
+                dm = sim_result.final_density_matrix
+                all_exp_vals.append([
+                    sum(
+                        term._expectation_from_density_matrix_no_validation(
+                            dm, qubit_order) for term in op) for op in o
+                ])
```

Comment:
I think this is doing unnecessary work and can be simplified to just:
```python3
for j, sub_op in enumerate(o):
  all_exp_vals[i][j] = sub_op.expectation_from_density_matrix_no_validation(dm, qubit_order)
```

Reply:
Done